File: sdk/ruby/boundary.rb
require 'json'

module Boundary
  class ValidationError < StandardError; end

  def self.validate(response, schema_class)
    # response can be Net::HTTPResponse or Faraday::Response
    status = response.respond_to?(:status) ? response.status : response.code.to_i
    body = response.body

    if status < 200 || status >= 300
      raise ValidationError, "Boundary: unexpected status code #{status}"
    end

    begin
      parsed = JSON.parse(body, symbolize_names: true)
      # In Ruby, we just instantiate the class with the parsed hash
      # Assumes schema_class has an initializer taking a hash
      schema_class.new(parsed)
    rescue JSON::ParserError => e
      raise ValidationError, "Boundary: schema validation failed (invalid JSON): #{e.message}"
    rescue => e
      raise ValidationError, "Boundary: schema validation failed: #{e.message}"
    end
  end
end

File: sdk/java/Boundary.java
package com.boundary.sdk;

import java.net.http.HttpResponse;
import com.fasterxml.jackson.databind.ObjectMapper;

public class Boundary {
    private static final ObjectMapper mapper = new ObjectMapper();

    public static <T> T validate(HttpResponse<String> response, Class<T> schemaClass) throws Exception {
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new RuntimeException("Boundary: unexpected status code " + response.statusCode());
        }

        String body = response.body();
        if (body == null || body.isEmpty()) {
            throw new RuntimeException("Boundary: empty response body");
        }

        try {
            return mapper.readValue(body, schemaClass);
        } catch (Exception e) {
            throw new RuntimeException("Boundary: schema validation failed", e);
        }
    }
}
